src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs

```diff
@@ -155,17 +171,18 @@ public void AddRange(IEnumerable<T> items)
 
                 while (enumerator.MoveNext())
                 {
-                    if (index == destination.Length)
+                    T item = enumerator.Current;
```

Comment:
If T is large, will this result in an extra copy that adds measurable cost?

Reply:
## Short answer

After investigation I don't see any problem with large T, and can't measure any noticable cost (outside of the noise / error).

Note: This PR is mainly about avoiding the range check in the common-path. So the larger T, the more dominant gets the cost for copying the value and the effort of the range check elimination is -> 0 (as can be seen in the benchmarks).

But while investigating this, I tried a way which slightly better codegen for the uncommon path, while the common path has equal codegen (variant _New2_, at the end of this answer).

## Long(er) answer

Let's break it down...with the following struct `Size64`:
```c#
public struct Size32
{
    public long A;
    public long B;
    public long C;
    public long D;
}

public struct Size64
{
    public Size32 A;
    public Size32 B;
}
```

Note: only the relevant dasm is shown!

### Current code

The current implementation results in:
```asm
G_M50766_IG04:
       45397E08             cmp      dword ptr [r14+8], r15d
       7531                 jne      SHORT G_M50766_IG08
       418BFF               mov      edi, r15d
       2B7B14               sub      edi, dword ptr [rbx+20]
       017B18               add      dword ptr [rbx+24], edi
       44897B14             mov      dword ptr [rbx+20], r15d
       488BFB               mov      rdi, rbx
       E86DF6FFFF           call     ...AllocateBuffer():this
       4C8B7308             mov      r14, gword ptr [rbx+8]
       448B7B14             mov      r15d, dword ptr [rbx+20]
       EB12                 jmp      SHORT G_M50766_IG08

G_M50766_IG05:
       458BE7               mov      r12d, r15d
       E980000000           jmp      G_M50766_IG09

G_M50766_IG06:
       E87EB39478           call     CORINFO_HELP_RNGCHKFAIL

G_M50766_IG07:
       458BFC               mov      r15d, r12d
       EBC9                 jmp      SHORT G_M50766_IG04

G_M50766_IG08:
       418D7701             lea      esi, [r15+1]
       448BE6               mov      r12d, esi
       488D75A0             lea      rsi, bword ptr [rbp-60H]
       488B7D98             mov      rdi, gword ptr [rbp-68H]
       49BB3000487C147F0000 mov      r11, 0x7F147C480030
       393F                 cmp      dword ptr [rdi], edi
       41FF13               call     qword ptr [r11]...:get_Current():struct:this
       453B7E08             cmp      r15d, dword ptr [r14+8]
       73D2                 jae      SHORT G_M50766_IG06
       4963FF               movsxd   rdi, r15d
       48C1E706             shl      rdi, 6
       498D7C3E10           lea      rdi, bword ptr [r14+rdi+16]
       C4E17A6F45A0         vmovdqu  xmm0, qword ptr [rbp-60H]              ; copy for destination[index++] = enumerator.Current;
       C4E17A7F07           vmovdqu  qword ptr [rdi], xmm0
       C4E17A6F45B0         vmovdqu  xmm0, qword ptr [rbp-50H]
       C4E17A7F4710         vmovdqu  qword ptr [rdi+16], xmm0
       C4E17A6F45C0         vmovdqu  xmm0, qword ptr [rbp-40H]
       C4E17A7F4720         vmovdqu  qword ptr [rdi+32], xmm0
       C4E17A6F45D0         vmovdqu  xmm0, qword ptr [rbp-30H]
       C4E17A7F4730         vmovdqu  qword ptr [rdi+48], xmm0
       488B7D98             mov      rdi, gword ptr [rbp-68H]
       49BB2800487C147F0000 mov      r11, 0x7F147C480028
       393F                 cmp      dword ptr [rdi], edi
       41FF13               call     qword ptr [r11]System.Collections.IEnumerator:MoveNext():bool:this
       85C0                 test     eax, eax
       7585                 jne      SHORT G_M50766_IG07
```

### Code of this PR (New1)

```asm
G_M5007_IG04:
       488D75A0             lea      rsi, bword ptr [rbp-60H]
       488B7D98             mov      rdi, gword ptr [rbp-68H]
       49BB5000498FE07F0000 mov      r11, 0x7FE08F490050
       393F                 cmp      dword ptr [rdi], edi
       41FF13               call     qword ptr [r11]...get_Current():struct:this
       488B55E8             mov      rdx, gword ptr [rbp-18H]
       8B5208               mov      edx, dword ptr [rdx+8]
       3955E0               cmp      dword ptr [rbp-20H], edx
       7245                 jb       SHORT G_M5007_IG05
       C4E17A6F45A0         vmovdqu  xmm0, qword ptr [rbp-60H]              ; copy for arg item in AddWithBufferAllocation
       C4E17A7F0424         vmovdqu  qword ptr [rsp], xmm0
       C4E17A6F45B0         vmovdqu  xmm0, qword ptr [rbp-50H]
       C4E17A7F442410       vmovdqu  qword ptr [rsp+10H], xmm0
       C4E17A6F45C0         vmovdqu  xmm0, qword ptr [rbp-40H]
       C4E17A7F442420       vmovdqu  qword ptr [rsp+20H], xmm0
       C4E17A6F45D0         vmovdqu  xmm0, qword ptr [rbp-30H]
       C4E17A7F442430       vmovdqu  qword ptr [rsp+30H], xmm0
       488D55E0             lea      rdx, bword ptr [rbp-20H]
       488D75E8             lea      rsi, bword ptr [rbp-18H]
       488BFB               mov      rdi, rbx
       E8F3FAFFFF           call     ...AddWithBufferAllocation(struct,byref,byref):this
       EB43                 jmp      SHORT G_M5007_IG06

G_M5007_IG05:
       488B7DE8             mov      rdi, gword ptr [rbp-18H]
       448B5DE0             mov      r11d, dword ptr [rbp-20H]
       4D63DB               movsxd   r11, r11d
       49C1E306             shl      r11, 6
       4A8D7C1F10           lea      rdi, bword ptr [rdi+r11+16]
       C4E17A6F45A0         vmovdqu  xmm0, qword ptr [rbp-60H]              ; copy for destination[index] = item
       C4E17A7F07           vmovdqu  qword ptr [rdi], xmm0
       C4E17A6F45B0         vmovdqu  xmm0, qword ptr [rbp-50H]
       C4E17A7F4710         vmovdqu  qword ptr [rdi+16], xmm0
       C4E17A6F45C0         vmovdqu  xmm0, qword ptr [rbp-40H]
       C4E17A7F4720         vmovdqu  qword ptr [rdi+32], xmm0
       C4E17A6F45D0         vmovdqu  xmm0, qword ptr [rbp-30H]
       C4E17A7F4730         vmovdqu  qword ptr [rdi+48], xmm0

G_M5007_IG06:
       8B7DE0               mov      edi, dword ptr [rbp-20H]
       FFC7                 inc      edi
       897DE0               mov      dword ptr [rbp-20H], edi
       488B7D98             mov      rdi, gword ptr [rbp-68H]
       49BB4800498FE07F0000 mov      r11, 0x7FE08F490048
       393F                 cmp      dword ptr [rdi], edi
       41FF13               call     qword ptr [r11]System.Collections.IEnumerator:MoveNext():bool:this
       85C0                 test     eax, eax
       0F8532FFFFFF         jne      G_M5007_IG04
```

So the copy for `destination[index] = item` is equal to the copy of the current code.

The copy for the argument for AddWithBufferAllocation can be "moved" to the call-site when using `private void AddWithBufferAllocation(in T item, ref T[] destination, ref int index)`.
This results in:
```asm
G_M5007_IG04:
       488D75A0             lea      rsi, bword ptr [rbp-60H]
       488B7D98             mov      rdi, gword ptr [rbp-68H]
       49BB50001E5F567F0000 mov      r11, 0x7F565F1E0050
       393F                 cmp      dword ptr [rdi], edi
       41FF13               call     qword ptr [r11]...get_Current():struct:this
       488B4DE8             mov      rcx, gword ptr [rbp-18H]
       8B4908               mov      ecx, dword ptr [rcx+8]
       394DE0               cmp      dword ptr [rbp-20H], ecx
       7216                 jb       SHORT G_M5007_IG05
       488D4DE0             lea      rcx, bword ptr [rbp-20H]
       488D75A0             lea      rsi, bword ptr [rbp-60H]
       488D55E8             lea      rdx, bword ptr [rbp-18H]
       488BFB               mov      rdi, rbx
       E828FBFFFF           call     ...AddWithBufferAllocation(byref,byref,byref):this
       EB43                 jmp      SHORT G_M5007_IG06

G_M5007_IG05:
       488B7DE8             mov      rdi, gword ptr [rbp-18H]
       448B5DE0             mov      r11d, dword ptr [rbp-20H]
       4D63DB               movsxd   r11, r11d
       49C1E306             shl      r11, 6
       4A8D7C1F10           lea      rdi, bword ptr [rdi+r11+16]
       C4E17A6F45A0         vmovdqu  xmm0, qword ptr [rbp-60H]
       C4E17A7F07           vmovdqu  qword ptr [rdi], xmm0
       C4E17A6F45B0         vmovdqu  xmm0, qword ptr [rbp-50H]
       C4E17A7F4710         vmovdqu  qword ptr [rdi+16], xmm0
       C4E17A6F45C0         vmovdqu  xmm0, qword ptr [rbp-40H]
       C4E17A7F4720         vmovdqu  qword ptr [rdi+32], xmm0
       C4E17A6F45D0         vmovdqu  xmm0, qword ptr [rbp-30H]
       C4E17A7F4730         vmovdqu  qword ptr [rdi+48], xmm0

G_M5007_IG06:
       8B7DE0               mov      edi, dword ptr [rbp-20H]
       FFC7                 inc      edi
       897DE0               mov      dword ptr [rbp-20H], edi
       488B7D98             mov      rdi, gword ptr [rbp-68H]
       49BB48001E5F567F0000 mov      r11, 0x7F565F1E0048
       393F                 cmp      dword ptr [rdi], edi
       41FF13               call     qword ptr [r11]System.Collections.IEnumerator:MoveNext():bool:this
       85C0                 test     eax, eax
       0F8561FFFFFF         jne      G_M5007_IG04
```
And in `AddWithBufferAllocation`:
```asm
       E82FFAFFFF           call     ...AllocateBuffer():this
       488B7308             mov      rsi, gword ptr [rbx+8]
       498BFC               mov      rdi, r12
       E83431AE78           call     CORINFO_HELP_CHECKED_ASSIGN_REF
       8B4314               mov      eax, dword ptr [rbx+20]
       418906               mov      dword ptr [r14], eax
       488B4308             mov      rax, gword ptr [rbx+8]
       418B3E               mov      edi, dword ptr [r14]
       C4C17A6F07           vmovdqu  xmm0, qword ptr [r15]
       C4E17A7F442408       vmovdqu  qword ptr [rsp+08H], xmm0
       C4C17A6F4710         vmovdqu  xmm0, qword ptr [r15+16]
       C4E17A7F442418       vmovdqu  qword ptr [rsp+18H], xmm0
       C4C17A6F4720         vmovdqu  xmm0, qword ptr [r15+32]
       C4E17A7F442428       vmovdqu  qword ptr [rsp+28H], xmm0
       C4C17A6F4730         vmovdqu  xmm0, qword ptr [r15+48]
       C4E17A7F442438       vmovdqu  qword ptr [rsp+38H], xmm0
       3B7808               cmp      edi, dword ptr [rax+8]
       734B                 jae      SHORT G_M6050_IG04
       4863FF               movsxd   rdi, edi
       48C1E706             shl      rdi, 6
       488D443810           lea      rax, bword ptr [rax+rdi+16]
       C4E17A6F442408       vmovdqu  xmm0, qword ptr [rsp+08H]
       C4E17A7F00           vmovdqu  qword ptr [rax], xmm0
       C4E17A6F442418       vmovdqu  xmm0, qword ptr [rsp+18H]
       C4E17A7F4010         vmovdqu  qword ptr [rax+16], xmm0
       C4E17A6F442428       vmovdqu  xmm0, qword ptr [rsp+28H]
       C4E17A7F4020         vmovdqu  qword ptr [rax+32], xmm0
       C4E17A6F442438       vmovdqu  xmm0, qword ptr [rsp+38H]
       C4E17A7F4030         vmovdqu  qword ptr [rax+48], xmm0
```
So the copy (ref -> value) is done here.

Based on just this information, I think it's preferable to use `in`, so that copying (of the argument) happens in the uncommon path, thus making the code for `AddRange` smaller.

### Variant New2

I checked a different variant too (New2):

```c#
public void AddRange(IEnumerable<T> items)
{
    Debug.Assert(items != null);

    using (IEnumerator<T> enumerator = items.GetEnumerator())
    {
        T[] destination = _current;
        int index = _index;

        // Continuously read in items from the enumerator, updating _count
        // and _index when we run out of space.

        while (enumerator.MoveNext())
        {
            if ((uint)index >= (uint)destination.Length)
            {
                AddWithBufferAllocation(enumerator, ref destination, ref index);
            }
            else
            {
                destination[index] = enumerator.Current;
            }

            index++;
        }

        // Final update to _count and _index.
        _count += index - _index;
        _index = index;
    }
}

// Non-inline to improve code quality as uncommon path
[MethodImpl(MethodImplOptions.NoInlining)]
private void AddWithBufferAllocation(IEnumerator<T> enumerator, ref T[] destination, ref int index)
{
    _count += index - _index;
    _index = index;
    AllocateBuffer();
    destination = _current;
    index = _index;
    _current[index] = enumerator.Current;
}
```
results in:
```asm
G_M38860_IG04:
       488B55E0             mov      rdx, gword ptr [rbp-20H]
       8B5208               mov      edx, dword ptr [rdx+8]
       3955D8               cmp      dword ptr [rbp-28H], edx
       7216                 jb       SHORT G_M38860_IG05
       488D55E0             lea      rdx, bword ptr [rbp-20H]
       488D4DD8             lea      rcx, bword ptr [rbp-28H]
       488BFB               mov      rdi, rbx
       488B7590             mov      rsi, gword ptr [rbp-70H]
       E8C3F8FFFF           call     ...AddWithBufferAllocation(ref,byref,byref):this
       EB5A                 jmp      SHORT G_M38860_IG06

G_M38860_IG05:
       4C8B75E0             mov      r14, gword ptr [rbp-20H]
       448B7DD8             mov      r15d, dword ptr [rbp-28H]
       488D7598             lea      rsi, bword ptr [rbp-68H]
       488B7D90             mov      rdi, gword ptr [rbp-70H]
       49BB70001E5F567F0000 mov      r11, 0x7F565F1E0070
       393F                 cmp      dword ptr [rdi], edi
       41FF13               call     qword ptr [r11]...get_Current():struct:this
       4963FF               movsxd   rdi, r15d
       48C1E706             shl      rdi, 6
       498D7C3E10           lea      rdi, bword ptr [r14+rdi+16]
       C4E17A6F4598         vmovdqu  xmm0, qword ptr [rbp-68H]
       C4E17A7F07           vmovdqu  qword ptr [rdi], xmm0
       C4E17A6F45A8         vmovdqu  xmm0, qword ptr [rbp-58H]
       C4E17A7F4710         vmovdqu  qword ptr [rdi+16], xmm0
       C4E17A6F45B8         vmovdqu  xmm0, qword ptr [rbp-48H]
       C4E17A7F4720         vmovdqu  qword ptr [rdi+32], xmm0
       C4E17A6F45C8         vmovdqu  xmm0, qword ptr [rbp-38H]
       C4E17A7F4730         vmovdqu  qword ptr [rdi+48], xmm0

G_M38860_IG06:
       8B7DD8               mov      edi, dword ptr [rbp-28H]
       FFC7                 inc      edi
       897DD8               mov      dword ptr [rbp-28H], edi
       488B7D90             mov      rdi, gword ptr [rbp-70H]
       49BB68001E5F567F0000 mov      r11, 0x7F565F1E0068
       393F                 cmp      dword ptr [rdi], edi
       41FF13               call     qword ptr [r11]System.Collections.IEnumerator:MoveNext():bool:this
       85C0                 test     eax, eax
       0F8561FFFFFF         jne      G_M38860_IG04
```
and for `AddWithBufferAllocation`:
```asm
       E8A9F7FFFF           call     ...AllocateBuffer():this
       488B7308             mov      rsi, gword ptr [rbx+8]
       498BFC               mov      rdi, r12
       E83E22AE78           call     CORINFO_HELP_CHECKED_ASSIGN_REF
       8B7314               mov      esi, dword ptr [rbx+20]
       418936               mov      dword ptr [r14], esi
       488B5B08             mov      rbx, gword ptr [rbx+8]
       458B36               mov      r14d, dword ptr [r14]
       488D75A0             lea      rsi, bword ptr [rbp-60H]
       498BFF               mov      rdi, r15
       49BB80001E5F567F0000 mov      r11, 0x7F565F1E0080
       393F                 cmp      dword ptr [rdi], edi
       41FF13               call     qword ptr [r11]...get_Current():struct:this
       443B7308             cmp      r14d, dword ptr [rbx+8]
       7348                 jae      SHORT G_M27297_IG04
       4963C6               movsxd   rax, r14d
       48C1E006             shl      rax, 6
       488D440310           lea      rax, bword ptr [rbx+rax+16]
       C4E17A6F45A0         vmovdqu  xmm0, qword ptr [rbp-60H]
       C4E17A7F00           vmovdqu  qword ptr [rax], xmm0
       C4E17A6F45B0         vmovdqu  xmm0, qword ptr [rbp-50H]
       C4E17A7F4010         vmovdqu  qword ptr [rax+16], xmm0
       C4E17A6F45C0         vmovdqu  xmm0, qword ptr [rbp-40H]
       C4E17A7F4020         vmovdqu  qword ptr [rax+32], xmm0
       C4E17A6F45D0         vmovdqu  xmm0, qword ptr [rbp-30H]
       C4E17A7F4030         vmovdqu  qword ptr [rax+48], xmm0
```

## Benchmarks

[Code for benchmarks](https://github.com/gfoidl/Benchmarks/blob/2b7a92c27c7f333cf2825d3bc1632c05890021d4/corefx/System/Collections/Generic/LargeArrayBuilder/source/LargeArrayBuilderBenchmarks/Benchmarks/AddRangeTSizeBenchmarks.cs)
The _LargeArrayBuilders_ used in the benchmars are [here](https://github.com/gfoidl/Benchmarks/tree/2b7a92c27c7f333cf2825d3bc1632c05890021d4/corefx/System/Collections/Generic/LargeArrayBuilder/source/LargeArrayBuilderBenchmarks).

Unfortunately I deleted the markdown file with the results and kept only the [HTML file](https://github.com/gfoidl/Benchmarks/blob/2b7a92c27c7f333cf2825d3bc1632c05890021d4/corefx/System/Collections/Generic/LargeArrayBuilder/source/LargeArrayBuilderBenchmarks/results/AddRangeTSizeBenchmarks/AddRangeTSizeBenchmarks.html).

Note: _New1_ uses `in` in `AddWithBufferAllocation`

## Question

Shall I update the PR to use the variant _New2_?

Reply:
> Shall I update the PR to use the variant New2?

Why don't we just go with what's in your PR then and you can follow-up on New2 if you think there's value there, but what's here seems fine.

Thanks.